### PR TITLE
agent/grpc: fix some test flakes and handle duplicate server IDs in the pool

### PR DESCRIPTION
**Pull request by dnephin**
This PR does a bunch of cleanup in `agent/grpc`, primarily in the tests. Best viewed by individual commit.

Adds a test, and fixes most of the test flakes (see comment below).

### CODE REVIEW

File: agent/grpc/client.go [hunk 1]
```diff
@@ -61,8 +61,7 @@ func (c *ClientConnPool) ClientConn(datacenter string) (*grpc.ClientConn, error)
 		grpc.WithInsecure(),
 		grpc.WithContextDialer(c.dialer),
 		grpc.WithDisableRetry(),
-		// TODO: previously this statsHandler was shared with the Handler. Is that necessary?
```
Comment (dnephin) [inline agent/grpc/client.go:64]: I'm pretty sure that using separate statsHandlers is correct. We don't want to share a single counter for both client and server connections because those are emitted as separate metrics.

File: agent/grpc/stats_test.go [hunk 2]
```diff
@@ -64,23 +65,42 @@ func TestHandler_EmitsStats(t *testing.T) {
...
-var fastRetry = &retry.Timer{Timeout: 7 * time.Second, Wait: 2 * time.Millisecond}
+func assertDeepEqual(t *testing.T, x, y interface{}, opts ...cmp.Option) {
+	t.Helper()
+	if diff := cmp.Diff(x, y, opts...); diff != "" {
+		t.Fatalf("assertion failed: values are not equal\n--- expected\n+++ actual\n%v", diff)
+	}
+}
 
-func patchGlobalMetrics(t *testing.T) *fakeMetricsSink {
+func patchGlobalMetrics(t *testing.T) (*fakeMetricsSink, func()) {
```
Comment (dnephin) [inline agent/grpc/stats_test.go:103]: I tried having this return a `reset` function instead of using `t.Cleanup` so that I could reset immediately after the `statsHandler` is created, but it did not really fix the flake.

File: agent/grpc/server_test.go [hunk 3]
```diff
@@ -116,11 +141,36 @@ func (f *fakeRPCListener) handleConn(conn net.Conn) {
...
+	case pool.RPCGRPC:
 		f.handler.Handle(conn)
 		return
+
+	case pool.RPCTLS:
+		// occasionally we see a test client connecting to an rpc listener that
+		// was created as part of another test, despite none of the tests running
+		// in parallel.
+		// Maybe some strange grpc behaviour? I'm not sure.
+		if f.tlsConf == nil {
```
Comment (dnephin) [inline agent/grpc/server_test.go:155]: This is related to the comment below in `TestHandler_EmitsStats`

File: agent/grpc/stats_test.go [hunk 4]
```diff
@@ -64,23 +65,42 @@ func TestHandler_EmitsStats(t *testing.T) {
 	// Wait for the server to stop so that active_streams is predictable.
 	require.NoError(t, g.Wait())
 
+	// Occasionally the active_stream=0 metric may be emitted before the
+	// active_conns=0 metric. The order of those metrics is not really important
+	// so we sort the calls to match the expected.
```
Comment (dnephin) [inline agent/grpc/stats_test.go:70]: This handling fixes one flake, but there is still one left:
```
--- FAIL: TestHandler_EmitsStats (0.00s)
    stats_test.go:88: assertion failed: values are not equal
        --- expected
        +++ actual
          []grpc.metricCall{
                {key: {"testing", "grpc", "server", "active_conns"}, val: 1},
        -       {key: []string{"testing", "grpc", "server", "active_streams"}, val: 1},
                {key: {"testing", "grpc", "server", "active_conns"}},
        +       {key: []string{"testing", "grpc", "server", "active_conns"}, val: 1},
        +       {key: []string{"testing", "grpc", "server", "active_conns"}, val: 2},
        +       {key: []string{"testing", "grpc", "server", "active_conns"}, val: 1},
        +       {key: []string{"testing", "grpc", "server", "active_conns"}},
        +       {key: []string{"testing", "grpc", "server", "active_streams"}, val: 1},
                {key: {"testing", "grpc", "server", "active_streams"}},
          }
```

I haven't been able to locate the problem. If I run this test by itself (`go test -count=1000 -run TestHandler_EmitsStats`) it never flakes, but if I run all the tests in the package it will occasionally fail because there are extra connections made.

From what I can tell (based on this flake, and the comment above in the fakeRPCListener) these servers are outliving the test, and a client from another test can connect to them. I'm not sure how that is possible because we use a unique scheme for each test, and we wait for the server to shutdown in `t.Cleanup`.

Maybe grpc is doing something I don't understand in background goroutines? Or maybe `t.Cleanup` is still running when the next test is started?

It takes a few hundred runs to reproduce the flake, so hopefully we don't see them very often in CI.

File: agent/grpc/handler.go [hunk 5]
```diff
@@ -75,7 +75,7 @@ func (l *chanListener) Addr() net.Addr {
 }
 
 func (l *chanListener) Close() error {
-	close(l.conns)
+	close(l.done)
```
Comment (banks) [inline agent/grpc/handler.go:78]: I don't think it's bad this way, but was there a specific problem with relying on closing the conns chan as a stop signal rather than a separate explicit done chan?

File: agent/grpc/resolver/resolver.go [hunk 6]
```diff
@@ -141,12 +129,16 @@ func (s *ServerResolverBuilder) AddServer(server *metadata.Server) {
 	}
 }
 
+func uniqueID(server *metadata.Server) string {
+	return server.Datacenter + "-" + server.ID
+}
```
Comment (banks) [inline agent/grpc/resolver/resolver.go:134]: Is it worth commenting the rationale for this function? The comment above maybe was too pedantic in that server node IDs kinda need to be unique across datacenters already IIRC to avoid issues with WAN gossip flapping since they are all in the same gossip pool. I don't specifically recall if it's a hard requirement though since raft/serf/catalog all treat node IDs a bit differently.

In the normal case they are all random UUIDs so are always unique anyway, it is just about possible that an operator could have chosen to override them all with `-node-id` and chosen to set them all to something the same etc.

In other words, I _think_ in all but the most bizarre edge case I can imagine, they are already globally unique in the cluster, I'm just not 100% sure that it's actually impossible to get things working with duplicated ones if you really try, it's certainly never advisable or reasonable too.

All that said, this is a simple way to ensure it's never not true so its fine to keep but it might puzzle future us as to why it's necessary so a comment might help?

File: agent/grpc/stats_test.go (hunk 4 above)
Comment (banks) [reply agent/grpc/stats_test.go:70]: Good detective work! Sounds like this is a good point to leave it for now and we can invest more if we see it causing CI issues.

**Review by banks — APPROVED** [review]
LGTM! Thanks.

File: agent/grpc/handler.go (hunk 5 above)
Comment (dnephin) [reply agent/grpc/handler.go:78]: Ya, I isolated this change in 2d70d24218d631fb273ed45e60410eae27bea5e6:

Closing l.conns can lead to a race and a 'panic: send on closed chan' when a connection is in the middle of being handled when the server is shutting down.

Found using `-race -count=800`.

File: agent/grpc/resolver/resolver.go (hunk 6 above)
Comment (dnephin) [reply agent/grpc/resolver/resolver.go:134]: Good idea! I'll add a comment explaining the rational here
